src/types.jl

```diff
@@ -118,42 +56,134 @@ function Base.show(io::IO, key::PublicKey)
     print("PublicKey{", backend_name(key), "}()")
 end
 
-# Here `capacity` needs to be large enough to hold at least `prod(shape)` elements
-mutable struct SecureMatrix{CryptoBackendT <: AbstractCryptoBackend, DataT}
+"""
+    SecureArray{Backend, N, DataT}
+
+Holds an encrypted `N`-dimensional array for arithmetic operations.
+Can be converted to a `PlainArray` using [`decrypt`](@ref).
+
+See also: [`PlainArray`](@ref), [`decrypt`](@ref)
+"""
+struct SecureArray{CryptoBackendT <: AbstractCryptoBackend, N, DataT}
     data::DataT
-    shape::Tuple{Int, Int}
-    capacity::Int
+    shape::Tuple
```

Comment:
This is an incomplete type (and thus type unstable). You likely want something like this:
```suggestion
    shape::NTuple{N, Int}
```